WIP: Changes for OpenVR 1.0.7

Ralith opened the pull request:
This gets the library building against the proposed -sys update and adds support for some important additional functionality such as controller input events. However, it is incomplete; it derives from what was originally intended to be a ground-up rewrite, and focuses on the bare minimum necessary to get up and running.

Substantial refactoring--some necessary, some cleanup, and some cosmetic--is included, again as a result of the original conception as a rewrite.

palodequeso commented:
Looks like this depends on Vulkan now? I'm not a maintainer/reviewer, but I was just reading through. I plan to make a VR game with Amethyst and this at some point. Note: I understand that you pretty much need a Vulkan compatible device to run VR anyway, I'm just curious.

reneeichhorn commented on src/lib.rs:
```diff
@@ -1,62 +1,143 @@
+/// Entry points into OpenVR.
+///
+/// At most one of this object may exist at a time.
+pub struct Context {
```
What I don't like on this approach is that you are forcing the user to instantiate all interfaces / subsystems.
If one of them fails the whole Context will be invalid. I think depending on the ApplicationType some interfaces are not available in will cause an error. The interfaces should be independent from each other.
As a quick solution we'd at least wrap an Option around each interface and make them all optional.

reneeichhorn commented:
@palodequeso The vulkan code you mentioned is actually part of OpenVR itself.
See https://github.com/ValveSoftware/openvr/wiki/Vulkan

reneeichhorn commented on src/compositor.rs:
```diff
@@ -1,73 +1,199 @@
+                Vulkan(ref x) => x as *const _ as *mut _,
+            },
+            eType: match texture.handle {
+                Vulkan(_) => sys::ETextureType_ETextureType_TextureType_Vulkan,
```
Reminder: 
ToDo: OpenGL support

palodequeso commented:
Oh rad, that makes sense, since it does the eye compositing IIRC.

reneeichhorn commented on src/compositor.rs:
```diff
@@ -1,73 +1,199 @@
+pub use sys::VkQueue_T;
+
+#[derive(Debug, Copy, Clone)]
+pub struct Texture {
```
Reminder:
ToDo: I'd prefer it to split the compositor into a module with multiple submodules for texture, etc.
```
+ compositor
    + mod.rs <- general logic
    + texture<- texture related stuff
           + mod.rs
           + vulkan.rs
           + opengl.rs
    + error.rs <- error enum ?maybe not?
```

reneeichhorn commented on src/lib.rs:
```diff
@@ -1,62 +1,143 @@
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum Eye {
```
I actually like that they were in the common module since they are also used in some other interfaces.

Ralith replied:
Since these are in the root, they can be used from any module in the crate.

reneeichhorn commented on src/system.rs:
```diff
@@ -1,112 +1,587 @@
     }
 }
+
+pub type PropertyContainerHandle = sys::PropertyContainerHandle_t;
```
We can definitely split this up

Ralith commented:
To be clear, Vulkan is not, in principle, a hard dep of OpenVR; the new code just does not expose the OpenGL/DX paths.

Ralith commented:
I split up the modules some more and drafted OpenGL support. I adjusted Context not to eagerly load subsystems; now it merely serves as a zero-cost safety mechanism.

As far as I know, github doesn't have any ability to target a PR at a nonexistent branch. If you'd like to create one, that might be helpful to allow this to be merged early and permit collaboration towards updating the remaining components.

reneeichhorn commented:
I changed the target branch to "development" (branched of from master)

reneeichhorn commented:
Merged it now into the development branch for further improvements and testing
